### PR TITLE
Typescript new elements

### PR DESCRIPTION
**Pull request by sorvell**
Adds
* TabBar and Tab (replaces previous version)
* Drawer
* TopAppBar

### CODE REVIEW

File: packages/drawer/src/mwc-drawer.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import {BaseElement, html, property, observer, query, customElement, Adapter, Foundation, PropertyValues} from '@material/mwc-base/base-element';
+import {classMap} from 'lit-html/directives/classMap.js';
+import MDCModalDrawerFoundation from '@material/drawer/modal/foundation.js';
+import MDCDismissibleDrawerFoundation from '@material/drawer/dismissible/foundation.js';
+import {strings} from '@material/drawer/constants.js';
+//import {MDCModalDrawerFoundation, MDCDismissibleDrawerFoundation, strings, util, createFocusTrap} from '@material/drawer/foundation.js';
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Remove comment

File: packages/drawer/src/mwc-drawer.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  new(adapter: Adapter): DrawerFoundation;
+}
+
+@customElement('mwc-drawer' as any)
+export class Drawer extends BaseElement {
+
+  @query('.mdc-drawer')
+  mdcRoot!: HTMLElement;
+
+  protected mdcFoundation: MDCDismissibleDrawerFoundation|MDCModalDrawerFoundation|undefined;
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: remove `undefined` and make this `mdcFoundation!:`

File: packages/drawer/src/mwc-drawer.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        this.dispatchEvent(new Event(strings.OPEN_EVENT, {bubbles: true, cancelable: true}))
+      },
+      trapFocus: () => {/*this._focusTrap.activate()*/},
+      releaseFocus: () => {/*this._focusTrap.deactivate()*/},
+    }
+  }
+
+  // TODO(sorvell): integrate focus trapping.
+  //private _focusTrap = undefined;
+  private _previousFocus: HTMLElement|undefined = undefined;
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Let's try to standardize on `| null`

File: packages/drawer/src/mwc-drawer.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  render() {
+    const dismissible = this.type === 'dismissible' || this.type === 'modal';
+    const modal = this.type === 'modal';
+    return html`
+      ${this.renderStyle()}
+      <aside class="mdc-drawer
+          ${classMap({'mdc-drawer--dismissible': dismissible, 'mdc-drawer--modal': modal})}">
+        ${this.hasHeader ? html`
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Can we have this pulled out and make the ternary simpler?

File: packages/drawer/src/mwc-drawer.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        <div class="mdc-drawer__content"><slot></slot></div>
+      </aside>
+      ${modal ? html`<div class="mdc-drawer-scrim" @click="${() => this.mdcFoundation.handleScrimClick()}"></div>` : ''}
+      `;
+  }
+
+  // note, we avoid calling `super.firstUpdated()` to control when `createFoundation()` is called.
+  firstUpdated() {
+    this.mdcRoot.addEventListener('keydown', (e) => {
+      if (this.mdcFoundation) {
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Remove check if `mdcFoundation` cannot be `undefined`

File: packages/drawer/src/mwc-drawer.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        ${this.hasHeader ? html`
+        <div class="mdc-drawer__header">
+          <h3 class="mdc-drawer__title"><slot name="title"></slot></h3>
+          <h6 class="mdc-drawer__subtitle"><slot name="subtitle"></slot></h6>
+          <slot name="header"></slot>
+        </div>
+        ` : ''}
+        <div class="mdc-drawer__content"><slot></slot></div>
+      </aside>
+      ${modal ? html`<div class="mdc-drawer-scrim" @click="${() => this.mdcFoundation.handleScrimClick()}"></div>` : ''}
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Refactor to a class field

File: packages/drawer/src/mwc-drawer.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  updated(changedProperties: PropertyValues) {
+    if (changedProperties.has('type')) {
+      this.createFoundation();
+    }
+  }
+
+  createFoundation() {
+    if (this.mdcFoundation !== undefined) {
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Let's move this logic in `BaseElement`

File: packages/drawer/src/mwc-drawer.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      this.mdcFoundation.destroy();
+    }
+    super.createFoundation();
+  }
+}
+
+declare global {
```
Comment (azakus) [inline packages/drawer/src/mwc-drawer.ts]: Standardize to above class

File: packages/tab-bar/src/mwc-tab-bar.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  private _getTab(index) {
+    return this._getTabs()[index];
+  }
+
+  createAdapter() {
+    return {
+      ...super.createAdapter(),
+      scrollTo: (scrollX) => this.scrollerElement.scrollToPosition(scrollX),
```
Comment (azakus) [inline packages/tab-bar/src/mwc-tab-bar.ts]: Put types on Adapter function arguments

File: packages/tab-bar/src/mwc-tab-bar.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        this.dispatchEvent(
+          new CustomEvent(MDCTabBarFoundation.strings.TAB_ACTIVATED_EVENT,
+            {detail: {index}, bubbles: true, cancelable: true}))
+      }
+    };
+  }
+
+  firstUpdated() {}
+
+  // NOTE: Delay creating foundation until scroller is fully updated.
```
Comment (azakus) [inline packages/tab-bar/src/mwc-tab-bar.ts]: Move note above `firstUpdated` override.

File: packages/tab-indicator/src/mwc-tab-indicator.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    if (changedProperties.has('fade')) {
+      this.createFoundation();
+    }
+  }
+
+  createAdapter() {
+    return {
+      ...super.createAdapter(),
+      computeContentClientRect: () => this.contentElement.getBoundingClientRect(),
+      setContentStyleProperty: (prop, value) => this.contentElement.style.setProperty(prop, value)
```
Comment (azakus) [inline packages/tab-indicator/src/mwc-tab-indicator.ts]: Types for Adapter function arguments

File: packages/tab-scroller/src/mwc-tab-scroller.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+  renderStyle() {
+    return style;
+  }
+
+  render() {
+    return html`
+      ${this.renderStyle()}
+      <div class="mdc-tab-scroller">
+        <div class="mdc-tab-scroller__scroll-area"
+            @wheel="${this._handleInteraction}"
```
Comment (azakus) [inline packages/tab-scroller/src/mwc-tab-scroller.ts]: `Wheel` event fires too often, remove or use `scroll` event.

File: packages/tab-scroller/src/mwc-tab-scroller.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+          <div class="mdc-tab-scroller__scroll-content"><slot></slot></div>
+        </div>
+      </div>
+      `;
+  }
+
+  createAdapter() {
+    return {
+      ...super.createAdapter(),
+      eventTargetMatchesSelector: (evtTarget, selector) => {
```
Comment (azakus) [inline packages/tab-scroller/src/mwc-tab-scroller.ts]: Types for Adapter Function args

File: packages/tab/src/mwc-tab.ts [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+  renderIndicator() {
+    return html`<mwc-tab-indicator
+        .icon="${this.indicatorIcon}"
+        .fade="${this.isFadingIndicator}"></mwc-tab-indicator>`;
+  }
+
+
+  createAdapter() {
+    return {
+      ...super.createAdapter(),
```
Comment (azakus) [inline packages/tab/src/mwc-tab.ts]: Types

File: packages/top-app-bar/src/mwc-top-app-bar.scss [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+:host {
+  display: inline-flex;
+}
+
+.mdc-top-app-bar {
+  flex: 1;
+}
+
+.mdc-top-app-bar--prominent {
+  /*.mdc-top-app-bar__action-item,
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.scss]: Remove comment

File: packages/top-app-bar/src/mwc-top-app-bar.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+  dense = false;
+
+  // does not work with prominent
+  @property({type: Boolean, reflect: true})
+  extraRow = false;
+
+  private _scrollTarget: HTMLElement|undefined;
+
+  get scrollTarget() {
+    return this._scrollTarget || window as any;
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.ts]: Does `window as Window` work?

File: packages/top-app-bar/src/mwc-top-app-bar.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+      <div class="mdc-top-app-bar__row">
+        <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start">
+          <slot name="navigationIcon"></slot>
+          <span class="mdc-top-app-bar__title"><slot name="title"></slot></span>
+        </section>
+        <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end" role="toolbar">
+          <slot name="actionItems"></slot>
+        </section>
+      </div>
+      ${this.extraRow ? html`
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.ts]: Pull multiline ternary up

File: packages/top-app-bar/src/mwc-top-app-bar.ts [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+      registerResizeHandler: (handler) => window.addEventListener('resize', handler),
+      deregisterResizeHandler: (handler) => window.removeEventListener('resize', handler),
+      getViewportScrollY: () => this.scrollTarget[this.scrollTarget === window ? 'pageYOffset' : 'scrollTop'],
+      getTotalActionItems: () =>
+        this._actionItemsSlot.assignedNodes({flatten: true}).length,
+    };
+  }
+
+  createFoundation() {
+    if (this.mdcFoundation) {
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.ts]: Remove, add to `BaseElement`

File: packages/top-app-bar/src/mwc-top-app-bar.ts [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    // update foundation if `type` or `scrollTarget` changes
+    if (changedProperties.has('type') || changedProperties.has('scrollTarget')) {
+      this.createFoundation();
+    }
+  }
+
+  protected createAdapter() {
+    return {
+      ...super.createAdapter(),
+      setStyle: (property, value) => this.mdcRoot.style.setProperty(property, value),
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.ts]: Types

File: packages/top-app-bar/src/mwc-top-app-bar.ts [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+    super.createFoundation();
+    const windowScroller = this.scrollTarget === window;
+    // we add support for top-app-bar's tied to an element scroller.
+    this.mdcRoot.style.position = windowScroller ? '' : 'absolute';
+    // TODO(sorvell): not sure why this is necessary but the MDC demo does it.
+    this.mdcRoot.style.top = windowScroller ? '0px' : '';
+  }
+}
+
+declare global {
```
Comment (azakus) [inline packages/top-app-bar/src/mwc-top-app-bar.ts]: Move above class

File: packages/drawer/src/mwc-drawer.ts (hunk 1 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 2 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 3 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 4 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 5 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done.

File: packages/drawer/src/mwc-drawer.ts (hunk 6 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 7 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/drawer/src/mwc-drawer.ts (hunk 8 above)
Comment (sorvell) [reply packages/drawer/src/mwc-drawer.ts]: Done

File: packages/tab-bar/src/mwc-tab-bar.ts (hunk 9 above)
Comment (sorvell) [reply packages/tab-bar/src/mwc-tab-bar.ts]: Done

File: packages/tab-bar/src/mwc-tab-bar.ts (hunk 10 above)
Comment (sorvell) [reply packages/tab-bar/src/mwc-tab-bar.ts]: Done

File: packages/tab-indicator/src/mwc-tab-indicator.ts (hunk 11 above)
Comment (sorvell) [reply packages/tab-indicator/src/mwc-tab-indicator.ts]: Done

File: packages/top-app-bar/src/mwc-top-app-bar.ts (hunk 16 above)
Comment (sorvell) [reply packages/top-app-bar/src/mwc-top-app-bar.ts]: Yup

File: packages/top-app-bar/src/mwc-top-app-bar.ts (hunk 17 above)
Comment (sorvell) [reply packages/top-app-bar/src/mwc-top-app-bar.ts]: Done

File: packages/top-app-bar/src/mwc-top-app-bar.ts (hunk 18 above)
Comment (sorvell) [reply packages/top-app-bar/src/mwc-top-app-bar.ts]: Done

File: packages/top-app-bar/src/mwc-top-app-bar.ts (hunk 19 above)
Comment (sorvell) [reply packages/top-app-bar/src/mwc-top-app-bar.ts]: Done

File: packages/top-app-bar/src/mwc-top-app-bar.ts (hunk 20 above)
Comment (sorvell) [reply packages/top-app-bar/src/mwc-top-app-bar.ts]: Done

File: packages/tab-bar/README.md [hunk 21]
```diff
@@ -0,0 +1,61 @@
+# mwc-tab-bar
+A [Material Components](https://material.io/components/) icon implementation using [Web Components](https://www.webcomponents.org/introduction)
```
Comment (kenchris) [inline packages/tab-bar/README.md]: this URL has changed to https://material.io/develop/

File: packages/tab-scroller/README.md [hunk 22]
```diff
@@ -0,0 +1,61 @@
+# mwc-tab-scroller
+A [Material Components](https://material.io/components/) icon implementation using [Web Components](https://www.webcomponents.org/introduction)
```
Comment (kenchris) [inline packages/tab-scroller/README.md]: s/icon/tab-scroller

File: packages/tab/README.md [hunk 23]
```diff
@@ -0,0 +1,61 @@
+# mwc-tab
+A [Material Components](https://material.io/components/) icon implementation using [Web Components](https://www.webcomponents.org/introduction)
```
Comment (kenchris) [inline packages/tab/README.md]: s/icon/mwc-tab

**Pull request merged**